core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java

```diff
@@ -99,30 +103,63 @@ public static long readLong(byte[] bytes, int start) {
    * @param <T> iterator value type
    * @return wrapped iterator
    */
+  // TODO(jiacheng): check all places this is used
   public static <T> CloseableIterator<T> createCloseableIterator(
       RocksIterator rocksIterator, RocksIteratorParser<T> parser) {
+    return createCloseableIterator(rocksIterator, parser, () -> false, () -> null);
+  }
+
+  /**
+   * Used to wrap an {@link CloseableIterator} over {@link RocksIterator}.
+   * It seeks given iterator to first entry before returning the iterator.
+   *
+   * closeCheck is used to stop the iterator before closing the underlying RocksDB.
+   * See RocksBlockMetaStore#getCloseableIterator for an example.
+   *
+   * @param rocksIterator the rocks iterator
+   * @param parser parser to produce iterated values from rocks key-value
+   * @param <T> iterator value type
+   * @param closeCheck if true, turn off the iteration
+   * @return wrapped iterator
+   */
+  public static <T> CloseableIterator<T> createCloseableIterator(
+      RocksIterator rocksIterator, RocksIteratorParser<T> parser,
+      Supplier<Boolean> closeCheck, Supplier<LockResource> locker) {
     rocksIterator.seekToFirst();
     AtomicBoolean valid = new AtomicBoolean(true);
     Iterator<T> iter = new Iterator<T>() {
       @Override
       public boolean hasNext() {
-        return valid.get() && rocksIterator.isValid();
+        /*
+         * There can be a race condition where after passing the hasNext() check, the RocksDB
+         * is closed and next() causes segfault. Avoiding that requires the closeCheck to stop the
+         * iteration way before entering the critical section. That means we can use the
+         * closeCheck to stop iteration when the iterator is still valid, then safely close
+         * RocksDB and all relevant references without worrying about concurrent readers
+         * like this escaped iterator.
+         */
+        return (!closeCheck.get()) && valid.get() && rocksIterator.isValid();
       }
 
       @Override
       public T next() {
-        try {
-          return parser.next(rocksIterator);
+        boolean succeeded = false;
+        // This lock will assure the RocksDB is not closed while this is reading
+        try (LockResource lock = locker.get()){
+          T result = parser.next(rocksIterator);
+          rocksIterator.next();
+          succeeded = true;
+          return result;
         } catch (Exception exc) {
```

Comment:
can we just catch throwable? it should cover exception

Reply:
the existing behavior is wrapping `Exception` into `RuntimeException`, i didn't want to change that